xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -949,18 +929,691 @@ func (s) TestRingHash_EndpointWeights(t *testing.T) {
 	if !cmp.Equal(got, want, cmpopts.EquateApprox(0, errorTolerance)) {
-		t.Errorf("Fraction of RPCs to backend %s: got %v, want %v (margin: +-%v)", backends[2].Address, got, want, errorTolerance)
+		t.Errorf("Fraction of RPCs to backend %s: got %v, want %v (margin: +-%v)", backends[2], got, want, errorTolerance)
+	}
+}
+
+// Tests that ring hash policy evaluation will continue past the terminal policy
+// if no results are produced yet.
+func (s) TestRingHash_ContinuesPastTerminalPolicyThatDoesNotProduceResult(t *testing.T) {
+	backends := startTestServiceBackends(t, 2)
+
+	const clusterName = "cluster"
+	endpoints := endpointResource(t, clusterName, backends)
+	cluster := e2e.ClusterResourceWithOptions(e2e.ClusterOptions{
+		ClusterName: clusterName,
+		ServiceName: clusterName,
+		Policy:      e2e.LoadBalancingPolicyRingHash,
+	})
+
+	route := e2e.DefaultRouteConfig("new_route", "test.server", clusterName)
+
+	// Even though this policy is terminal, since it produces no result, we
+	// continue past it to find a policy that produces results.
+	hashPolicy := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "header_not_present",
+			},
+		},
+	}
+	hashPolicy.Terminal = true
+	hashPolicy2 := v3routepb.RouteAction_HashPolicy{
+		PolicySpecifier: &v3routepb.RouteAction_HashPolicy_Header_{
+			Header: &v3routepb.RouteAction_HashPolicy_Header{
+				HeaderName: "address_hash",
+			},
+		},
+	}
+	action := route.VirtualHosts[0].Routes[0].Action.(*v3routepb.Route_Route)
+	action.Route.HashPolicy = []*v3routepb.RouteAction_HashPolicy{&hashPolicy, &hashPolicy2}
+
+	listener := e2e.DefaultClientListener(virtualHostName, route.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	xdsServer, nodeID, xdsResolver := setupManagementServerAndResolver(t)
+	if err := xdsServer.Update(ctx, xdsUpdateOpts(nodeID, endpoints, cluster, route, listener)); err != nil {
+		t.Fatalf("Failed to update xDS resources: %v", err)
+	}
+
+	conn, err := grpc.NewClient("xds:///test.server", grpc.WithResolvers(xdsResolver), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to create client: %s", err)
+	}
+	defer conn.Close()
+	client := testgrpc.NewTestServiceClient(conn)
+
+	// Note that the first policy does not match, and in the second hash policy,
+	// each type of RPC contains a header value that always hashes to a specific
+	// backend, as the header value matches the value used to create the entry
+	// in the ring.
+	wantBackend := backends[0]
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("address_hash", wantBackend+"_0"))
+	const numRPCs = 100
+	gotPerBackend := checkRPCSendOK(ctx, t, client, numRPCs)
+	if got := gotPerBackend[wantBackend]; got != numRPCs {
+		t.Errorf("Got %v RPCs routed to backend %v, want %v", got, wantBackend, numRPCs)
+	}
```

Comment:
I don't quite understand what is being tested here. I see that we configure two hash policies, both matching headers, with the first one containing the terminal field.

And the RPC is made with the header specified in the second hash policy. So, expect the second hash policy to be used. But how do we know that it was actually used? And how do we know that using the second hash policy leads to `backends[0]` being used? Maybe some comments would help.

Reply:
And if this is testing whether we go past terminal hash policies that don't generate a value, why is this being tested in the context of the ring_hash LB policy? Wouldn't a unit test in the component that is responsible for computing the hash value good enough?

Reply:
I modified the comment that hopefully clarifies what we are testing.

> And if this is testing whether we go past terminal hash policies that don't generate a value, why is this being tested in the context of the ring_hash LB policy? Wouldn't a unit test in the component that is responsible for computing the hash value good enough?

For context, the unit tests are here: https://github.com/grpc/grpc-go/blob/6e9c88b0acf15de2256c3d3a257869e0669572d9/xds/internal/resolver/serviceconfig_test.go#L62

I think there is an argument for doing both, as imo quite a bit can go wrong in the interactions between RDS parsing, properly setting the request hash based on hash policies, and applying the ring hash load balancing policy.

I admit that I ported the end to end tests from c-core that Mark originally wrote and added as bugs were uncovered, without necessarily challenging or carefully thinking about which case has the most chances of uncovering problems, but I don't think this one is clearly not appropriate to have.